File: Codeground/Codeground/parenthesis.c
/*
You should use the statndard input/output

in order to receive a score properly.

Do not use file input and output

Please be very careful.
*/

#include <stdio.h>
#define TRUE 1
#define FALSE 0

//str - 뒤에 이어지는 string, n - 남은 str의 개수
int check_s(char *str, int n) { 
	int i = 0;

	for (i = 0; i < n; i++)
	{
		if (str[i] == ')') 
		{
			return TRUE;
		}
	}

	return FALSE;
}

int check_m(char *str, int n) {
	int i = 0;

	for (i = 0; i < n; i++)
	{
		if (str[i] == '}')
		{
			return TRUE;
		}
	}

	return FALSE;
}

int check_l(char *str, int n) {
	int i = 0;

	for (i = 0; i < n; i++)
	{
		if (str[i] == ']')
		{
			return TRUE;
		}
	}

	return FALSE;
}

int Answer;

int main(void)
{
	int T, test_case;
	char *str;
	/*
	The freopen function below opens input.txt file in read only mode, and afterward,
	the program will read from input.txt file instead of standard(keyboard) input.
	To test your program, you may save input data in input.txt file,
	and use freopen function to read from the file when using scanf function.
	You may remove the comment symbols(//) in the below statement and use it.
	But before submission, you must remove the freopen function or rewrite comment symbols(//).
	*/
	int f = freopen("input.txt", "r", stdin);

	/*
	If you remove the statement below, your program's output may not be rocorded
	when your program is terminated after the time limit.
	For safety, please use setbuf(stdout, NULL); statement.
	*/
	setbuf(stdout, NULL);

	scanf("%d", &T);
	for (test_case = 0; test_case < T; test_case++)
	{
		Answer = 0;

		/////////////////////////////////////////////////////////////////////////////////////////////
		/*
		Implement your algorithm here.
		The answer to the case will be stored in variable Answer.
		*/
		/////////////////////////////////////////////////////////////////////////////////////////////

		scanf("%s", &str);

		// Print the answer to standard output(screen).

		printf("Case #%d\n", test_case + 1);
		printf("%d\n", Answer);

	}

	fclose(f);

	return 0;//Your program should return 0 on normal termination.
}

File: Codeground/Codeground_java/src/Parenthesis.java
//SCPC 3È¸ ¿¹¼± - °ýÈ£

import java.io.FileInputStream;
import java.util.*;

public class Parenthesis {

	static int Answer;
	static char[] list;
	static String str;
	static Stack st;
	static boolean[] matched;
	static int temp;
	

	public static void main(String args[]) throws Exception	{
		/*
		   The method below means that the program will read from input.txt, instead of standard(keyboard) input.
		   To test your program, you may save input data in input.txt file,
		   and call below method to read from the file when using nextInt() method.
		   You may remove the comment symbols(//) in the below statement and use it.
		   But before submission, you must remove the freopen function or rewrite comment symbols(//).
		 */		

		/*
		   Make new scanner from standard input System.in, and read data.
		 */
		Scanner sc = new Scanner(System.in);
		//Scanner sc = new Scanner(new FileInputStream("input.txt"));

		int T = sc.nextInt();
		
		for(int test_case = 0; test_case < T; test_case++) {

			Answer = 0;
			temp = 0;
			/////////////////////////////////////////////////////////////////////////////////////////////
			/*
			   Implement your algorithm here.
			   The answer to the case will be stored in variable Answer.
			 */
			/////////////////////////////////////////////////////////////////////////////////////////////

			str = sc.next();
			list = str.toCharArray();
			matched = new boolean[list.length];
			
			st = new Stack();
			//answerst = new Stack();
			
			Search(list);
			for(int i = 0; i < matched.length; i++){
				if(matched[i]){
					temp += 1;
					if(Answer < temp) Answer = temp;
				}
				else {
					temp = 0;
				}
			}
			
			// Print the answer to standard output(screen).
			System.out.println("Case #"+(test_case+1));
			System.out.println(Answer);
		}
	}
	
	public static void Search(char[] array){
		int pop;
		for(int i = 0; i < array.length; i++) {
			if(match(array[i]) == 'x'){
				if(st.isEmpty()){
					matched[i] = false;
				}
				else {
					pop = (int)st.pop();
					if(match(array[pop]) == array[i]){
						matched[pop] = true;
						matched[i] = true;
					}
					else {
						matched[i] = false;
						while(!st.isEmpty()){
							pop = (int)st.pop();
							matched[pop] = false;
						}
					}
				}
			}
			else {
				st.push(i);
				matched[i] = false;
			}
		}
	}
	
	public static char match(char c){
		if(c == '(') {
			return ')';
		}
		else if(c == '{') {
			return '}';
		}
		else if(c == '[') {
			return ']';
		}
		else return 'x';
	}
}